File: frontend/src/lib/preferences.ts
/**
 * 用户偏好管理 — K 线指标/形态选择
 *
 * 持久化优先级：
 * 1. 后端 Redis（X-User-Id 隔离，多端同步）— 走 PUT/GET /api/preferences
 * 2. localStorage 兜底（后端 503 时降级）
 *
 * anonymous user_id 在首次访问时生成并保存在 localStorage.uuid_user_id
 */

const USER_ID_KEY = "ai_trader_user_id";

export interface IndicatorPref {
  enabled: boolean;
  params: Record<string, number | string | boolean>;
}

export interface ChartPreferences {
  indicators: Record<string, IndicatorPref>;
}

const DEFAULT_PREFS: ChartPreferences = {
  indicators: {
    // 主图均线（用户重点要求多周期）
    ma5: { enabled: false, params: { length: 5 } },
    ma10: { enabled: false, params: { length: 10 } },
    ma20: { enabled: true, params: { length: 20 } },
    ma30: { enabled: true, params: { length: 30 } },
    ma60: { enabled: false, params: { length: 60 } },
    ema12: { enabled: false, params: { length: 12 } },
    ema26: { enabled: false, params: { length: 26 } },
    ema50: { enabled: true, params: { length: 50 } },
    // 完整布林带（3 条线）
    boll: { enabled: true, params: { length: 20, mult: 2.0 } },
    // VWAP
    vwap: { enabled: false, params: {} },
    // 副图指标
    rsi: { enabled: true, params: { length: 14 } },
    rsi6: { enabled: false, params: { length: 6 } },
    rsi24: { enabled: false, params: { length: 24 } },
    macd: { enabled: true, params: { fast: 12, slow: 26, signal: 9 } },
    kdj: { enabled: true, params: { length: 9 } },
    obv: { enabled: true, params: {} },
    stoch: { enabled: false, params: { k: 14, d: 3 } },
    cci: { enabled: false, params: { length: 20 } },
    wr: { enabled: false, params: { length: 14 } },
    mfi: { enabled: false, params: { length: 14 } },
    adx: { enabled: false, params: { length: 14 } },
    atr: { enabled: false, params: { length: 14 } },
    // 高级趋势
    sar: { enabled: false, params: { step: 0.02, max: 0.2 } },
    supertrend: { enabled: false, params: { length: 10, mult: 3.0 } },
    keltner: { enabled: false, params: { length: 20, mult: 2.0 } },
    ichimoku: { enabled: false, params: {} },
  },
};

function getUserId(): string {
  let uid = localStorage.getItem(USER_ID_KEY);
  if (!uid) {
    uid = `anon-${Date.now()}-${Math.random().toString(36).substring(2, 12)}`;
    localStorage.setItem(USER_ID_KEY, uid);
  }
  return uid;
}

const LOCAL_FALLBACK_KEY = "ai_trader_prefs_local";

function loadLocalFallback(): ChartPreferences {
  try {
    const raw = localStorage.getItem(LOCAL_FALLBACK_KEY);
    if (raw) return JSON.parse(raw) as ChartPreferences;
  } catch {
    /* ignore */
  }
  return DEFAULT_PREFS;
}

function saveLocalFallback(prefs: ChartPreferences): void {
  try {
    localStorage.setItem(LOCAL_FALLBACK_KEY, JSON.stringify(prefs));
  } catch {
    /* ignore */
  }
}

export async function loadPreferences(): Promise<ChartPreferences> {
  const uid = getUserId();
  try {
    const res = await fetch("/api/preferences", {
      headers: { "X-User-Id": uid },
    });
    if (!res.ok) throw new Error(`HTTP ${res.status}`);
    const data = await res.json();
    return data.preferences as ChartPreferences;
  } catch {
    // 后端不可用时回退到 localStorage
    return loadLocalFallback();
  }
}

export async function savePreferences(prefs: ChartPreferences): Promise<boolean> {
  const uid = getUserId();
  // 永远保存到 localStorage 兜底
  saveLocalFallback(prefs);
  try {
    const res = await fetch("/api/preferences", {
      method: "PUT",
      headers: { "Content-Type": "application/json", "X-User-Id": uid },
      body: JSON.stringify(prefs),
    });
    return res.ok;
  } catch {
    return false;
  }
}

export async function resetPreferences(): Promise<ChartPreferences> {
  const uid = getUserId();
  try {
    const res = await fetch("/api/preferences/reset", {
      method: "POST",
      headers: { "X-User-Id": uid },
    });
    if (!res.ok) throw new Error(`HTTP ${res.status}`);
    const data = await res.json();
    saveLocalFallback(data.preferences);
    return data.preferences;
  } catch {
    saveLocalFallback(DEFAULT_PREFS);
    return DEFAULT_PREFS;
  }
}

export function getUserIdForDebug(): string {
  return getUserId();
}

File: backend/app/api/preferences.py
"""User preferences 端点 — K 线指标/形态等用户选择（Redis 持久化）

按 anonymous user_id（前端生成 UUID，存在 localStorage）存储。
跨设备同步：用户只需在同一浏览器保留 UUID 即可。如未来接入正式用户系统，
把 anonymous_user_id 换成 user_id 即可，schema 不变。
"""

import json
import logging
from typing import Any

from fastapi import APIRouter, Header, HTTPException
from pydantic import BaseModel, Field

from app.data import binance_client

logger = logging.getLogger(__name__)

router = APIRouter(redirect_slashes=False)


# ── Schemas ──────────────────────────────────────────────────────────────────


class IndicatorPref(BaseModel):
    """单个指标的开关 + 可选参数覆盖"""

    enabled: bool = True
    params: dict[str, Any] = Field(default_factory=dict)


class PatternPref(BaseModel):
    """K 线形态标记开关"""

    enabled: bool = True


class ChartPreferences(BaseModel):
    """K 线图用户偏好"""

    indicators: dict[str, IndicatorPref] = Field(default_factory=dict)
    patterns: dict[str, PatternPref] = Field(default_factory=dict)


class PreferencesResponse(BaseModel):
    user_id: str
    preferences: ChartPreferences


# ── Helpers ──────────────────────────────────────────────────────────────────


def _key(user_id: str) -> str:
    return f"user_prefs:{user_id}"


async def _get_redis() -> Any | None:
    """复用 binance_client 的 redis 连接，避免重复创建"""
    return binance_client._redis  # noqa: SLF001 — intentional reuse


def _default_preferences() -> ChartPreferences:
    """新用户的默认偏好：保留项目核心指标 + 5 个最常用形态"""
    return ChartPreferences(
        indicators={
            # 主图均线（用户重点要求多周期）
            "ma5": IndicatorPref(enabled=False, params={"length": 5}),
            "ma10": IndicatorPref(enabled=False, params={"length": 10}),
            "ma20": IndicatorPref(enabled=True, params={"length": 20}),
            "ma30": IndicatorPref(enabled=True, params={"length": 30}),
            "ma60": IndicatorPref(enabled=False, params={"length": 60}),
            "ema12": IndicatorPref(enabled=False, params={"length": 12}),
            "ema26": IndicatorPref(enabled=False, params={"length": 26}),
            "ema50": IndicatorPref(enabled=True, params={"length": 50}),
            # 完整布林带（3 条线）
            "boll": IndicatorPref(enabled=True, params={"length": 20, "mult": 2.0}),
            # VWAP
            "vwap": IndicatorPref(enabled=False, params={}),
            # 副图指标
            "rsi": IndicatorPref(enabled=True, params={"length": 14}),
            "rsi6": IndicatorPref(enabled=False, params={"length": 6}),
            "rsi24": IndicatorPref(enabled=False, params={"length": 24}),
            "macd": IndicatorPref(enabled=True, params={"fast": 12, "slow": 26, "signal": 9}),
            "kdj": IndicatorPref(enabled=True, params={"length": 9}),
            "obv": IndicatorPref(enabled=True, params={}),
            "stoch": IndicatorPref(enabled=False, params={"k": 14, "d": 3}),
            "cci": IndicatorPref(enabled=False, params={"length": 20}),
            "wr": IndicatorPref(enabled=False, params={"length": 14}),
            "mfi": IndicatorPref(enabled=False, params={"length": 14}),
            "adx": IndicatorPref(enabled=False, params={"length": 14}),
            "atr": IndicatorPref(enabled=False, params={"length": 14}),
            # 高级趋势
            "sar": IndicatorPref(enabled=False, params={"step": 0.02, "max": 0.2}),
            "supertrend": IndicatorPref(enabled=False, params={"length": 10, "mult": 3.0}),
            "keltner": IndicatorPref(enabled=False, params={"length": 20, "mult": 2.0}),
            "ichimoku": IndicatorPref(enabled=False, params={}),
        },
        patterns={
            # Top 10 形态（按使用频率）
            "hammer": PatternPref(enabled=True),
            "hanging_man": PatternPref(enabled=True),
            "engulfing": PatternPref(enabled=True),
            "doji": PatternPref(enabled=True),
            "shooting_star": PatternPref(enabled=True),
            "morning_star": PatternPref(enabled=True),
            "evening_star": PatternPref(enabled=True),
            "three_white_soldiers": PatternPref(enabled=True),
            "three_black_crows": PatternPref(enabled=True),
            "harami": PatternPref(enabled=False),
        },
    )


# ── Endpoints ────────────────────────────────────────────────────────────────


@router.get("", response_model=PreferencesResponse)
async def get_preferences(
    x_user_id: str = Header(..., alias="X-User-Id", description="anonymous UUID from localStorage"),
) -> PreferencesResponse:
    """读取用户偏好。无记录时返回默认。"""
    redis = await _get_redis()
    if redis:
        try:
            raw = await redis.get(_key(x_user_id))
            if raw:
                data = json.loads(raw)
                return PreferencesResponse(user_id=x_user_id, preferences=ChartPreferences(**data))
        except Exception as e:
            logger.warning("preferences read failed: %s", e)

    return PreferencesResponse(user_id=x_user_id, preferences=_default_preferences())


@router.put("", response_model=PreferencesResponse)
async def put_preferences(
    body: ChartPreferences,
    x_user_id: str = Header(..., alias="X-User-Id"),
) -> PreferencesResponse:
    """整体覆盖保存用户偏好。"""
    redis = await _get_redis()
    payload = body.model_dump()
    if redis:
        try:
            await redis.set(_key(x_user_id), json.dumps(payload))
        except Exception as e:
            logger.warning("preferences write failed: %s", e)
            raise HTTPException(status_code=503, detail="cache unavailable")
    else:
        # Redis 不可用时直接返回（不报错），让前端保持 localStorage 兜底
        logger.warning("Redis unavailable — preferences not persisted for user %s", x_user_id)
    return PreferencesResponse(user_id=x_user_id, preferences=body)


@router.post("/reset", response_model=PreferencesResponse)
async def reset_preferences(
    x_user_id: str = Header(..., alias="X-User-Id"),
) -> PreferencesResponse:
    """重置为默认偏好"""
    redis = await _get_redis()
    if redis:
        try:
            await redis.delete(_key(x_user_id))
        except Exception as e:
            logger.warning("preferences delete failed: %s", e)
    return PreferencesResponse(user_id=x_user_id, preferences=_default_preferences())
